flocker/volume/filesystems/zfs.py

```diff
@@ -289,33 +341,46 @@ def writer(self):
             # a hack.  When we replace this mechanism with a proper API we
             # should make it include that information.
             #
-            # -e means "if the stream says it is for foo/bar/baz then receive
-            # into baz".  I don't know why self.name is also required,
-            # then. XXX try -d self.pool instead. XXX it works without -e w/
-            # self.name too. XXX Delete this paragraph if we go ahead with just
-            # `-F` in the implementation.
-            #
-            # -F means force.  If the stream is based on not-quite-the-latest
+            # If the stream is based on not-quite-the-latest
             # snapshot then we have to throw away all the snapshots newer than
             # it in order to receive the stream.  To do that you have to
             # force.
             #
-            cmd = [b"zfs", b"receive", b"-F", self.name]
+            force = True
         else:
             # If the filesystem doesn't already exist then this is a complete
             # data stream.
-            cmd = [b"zfs", b"receive", self.name]
-        process = Popen(cmd, stdin=PIPE)
-        succeeded = False
+            force = False
+
+        (rfd, wfd) = os.pipe()
+        wfile = os.fdopen(wfd, "w")
+        queue = Queue()
+
+        def recv_and_close():
+            try:
+                libzfs_core.lzc_receive(self.name, rfd, force)
+                success = True
+            except Exception:
```

Comment:
Looking at this again and it looks like `IFilesystem` does not specify any exception that could be raised in this case.
Also, the only other implementer of the interface, in-memory `DirectoryFilesystem`, does not raise any excepion either. More importantly, none of the functional tests expect an exception from `reader` or `writer` context managers.

So, I think that the refinement of the interface should be a separate issue.
@itamarst, do you agree?

What I can do right now is log the exception. Should that be a condition for merging this PR? Or could that be a separate issue for later?
If I do that now, can I re-use `ZFS_ERROR` message type here?